src/raftstore/store/worker/apply.rs

```diff
@@ -1105,6 +1151,270 @@ impl ApplyDelegate {
+        let mut merging_state = MergeState::new();
+        merging_state.set_min_index(index);
+        merging_state.set_target(prepare_merge.get_target().to_owned());
+        merging_state.set_commit(exec_ctx.index);
+        write_merge_state(
+            &self.engine,
+            ctx.wb(),
+            &region,
+            PeerState::Merging,
+            merging_state.clone(),
+        ).unwrap_or_else(|e| {
+            panic!(
+                "{} failed to save merging state {:?} for region {:?}: {:?}",
+                self.tag, merging_state, region, e
+            )
+        });
+
+        PEER_ADMIN_CMD_COUNTER_VEC
+            .with_label_values(&["prepare_merge", "success"])
+            .inc();
+
+        Ok((
+            AdminResponse::new(),
+            Some(ExecResult::PrepareMerge {
+                region: region,
+                state: merging_state,
+            }),
+        ))
+    }
+
+    fn load_entries_for_merge(&self, merge: &CommitMergeRequest, apply_index: u64) -> Vec<Entry> {
+        // Entries from [first_index, last_index) need to be loaded.
+        let first_index = apply_index + 1;
+        let last_index = merge.get_commit();
+        if first_index >= last_index {
+            return vec![];
+        }
+        let exist_first_index = merge
+            .get_entries()
+            .get(0)
+            .map_or(last_index, |e| e.get_index());
+        if first_index >= exist_first_index {
+            return merge.get_entries()[(first_index - exist_first_index) as usize..].to_vec();
+        }
+        let source_region = merge.get_source();
+        let mut entries = Vec::with_capacity((last_index - first_index) as usize);
+        peer_storage::fetch_entries_to(
+            &self.raft_engine,
+            source_region.get_id(),
+            first_index,
+            exist_first_index,
+            NO_LIMIT,
+            &mut entries,
+        ).unwrap_or_else(|e| {
+            panic!(
+                "{} failed to load entries [{}:{}) from region {}: {:?}",
+                self.tag,
+                first_index,
+                exist_first_index,
+                source_region.get_id(),
+                e
+            );
+        });
+        entries.extend_from_slice(merge.get_entries());
+        entries
+    }
+
+    fn catch_up_log_for_merge(
+        &mut self,
+        ctx: &mut ApplyContext,
+        merge: &CommitMergeRequest,
+        exist_region: &Region,
+    ) {
+        let source_region = merge.get_source();
+        let apply_state_key = keys::apply_state_key(source_region.get_id());
+        let apply_state: RaftApplyState = match self.engine.get_msg_cf(CF_RAFT, &apply_state_key) {
+            Ok(Some(s)) => s,
+            e => panic!(
+                "{} failed to get apply state of {:?}: {:?}",
+                self.tag, source_region, e
+            ),
+        };
+        let apply_index = apply_state.get_applied_index();
```

Comment:
Why not using the index of last log? Is it possible that apply_index < min_index?

Reply:
Because the function tries to catch up logs for source region, so it needs to apply logs from last apply position.
It's possible. That's one of the cases function `load_entries_for_merge` tries to solve.